### PR TITLE
Converted FullNodeConfiguration into a data class and added ability t…

### PR DESCRIPTION
**Pull request by shamsasari**
…o parse Configs into data classes

This is some clean up work which I started some time back and hadn't got round to finishing it. The idea is to remove the boilerplate involved in adding new config options to FullNodeConfiguration by making it into a data class and parsing that. Typesafe Config already has the ability to work with JavaBeans (i.e. no args c'tor, setters, getters, etc) which data classes aren't.

I've also added in an `OldConfig` annotation to enable migration of code config to new ones, for example if a property needs to be renamed. For now it just logs a warning if the old config is used. This was needed because parsing FullNodeConfiguration as a data class mean the RPC `User` objects required a `username` field rather than the existing `user` one. I could have renamed `username` to `user` but actually `username` is a slightly more descriptive config key.

I also changed up the `certificateChainCheckPolicies` config a bit to make it easier to parse and also easier to configure. This config isn't actually documented anywhere so this shouldn't be a problem.

### CODE REVIEW

**Comment by shamsasari** [comment]
I had to remove `webAddress` from FullNodeConfiguration as the the trader demo nodes weren't starting due to it being missing. It seems the old way of parsing was lazy and this is now eager (which is good IMO). webAddress isn't used by the node anyway so it makes sense for it not to be in FullNodeConfiguration. I've updated the driver so it works - it's been added to NodeHandle.

File: node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt [hunk 1]
```diff
@@ -12,103 +11,103 @@ import net.corda.node.services.messaging.CertificateChainCheckPolicy
...
 
-private fun Config.getOptionalConfig(path: String): Config? = if (hasPath(path)) getConfig(path) else null
+enum class CertChainPolicyType {
+    Any,
+    RootMustMatch,
+    LeafMustMatch,
+    MustContainOneOf
+}
+
+data class CertChainPolicyConfig(val role: String, val policy: CertChainPolicyType, val trustedAliases: Set<String>) {
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt]: Hah! Was wondering how you'll do this one:)

File: node-api/src/main/kotlin/net/corda/nodeapi/config/ConfigUtilities.kt [hunk 2]
```diff
@@ -2,73 +2,126 @@ package net.corda.nodeapi.config
...
+private fun <T> Config.getValueInternal(path: String, type: KType): T {
+    return (if (type.arguments.isEmpty()) getSingleValue(path, type) else getCollectionValue(path, type)) as T
+}
+
+private fun Config.getSingleValue(path: String, type: KType): Any? {
+    if (type.isMarkedNullable && !hasPath(path)) return null
+    val typeClass = type.jvmErasure
+    return when (typeClass) {
+        String::class -> getString(path)
+        Int::class -> getInt(path)
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/config/ConfigUtilities.kt]: That's what I originally had when I was using Java reflection. But using Kotlin's reflection API avoids that, which is nice.

File: node-api/src/main/kotlin/net/corda/nodeapi/config/ConfigUtilities.kt [hunk 3]
```diff
@@ -2,73 +2,126 @@ package net.corda.nodeapi.config
...
 @Suppress("UNCHECKED_CAST")
-inline fun <reified T : Any> Config.getListOrElse(path: String, default: Config.() -> List<T>): List<T> {
-    return if (hasPath(path)) {
-        (if (T::class == String::class) getStringList(path) else getConfigList(path)) as List<T>
-    } else {
-        this.default()
-    }
+private fun parseEnum(enumType: Class<*>, name: String): Enum<*> {
+    val values = enumType.getMethod("values").invoke(null) as Array<Enum<*>>
+    return values.firstOrNull { it.name == name } ?: throw IllegalArgumentException("$name is not a value of $enumType")
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/config/ConfigUtilities.kt]: I tried using your `DummyEnum` trick but I couldn't get it to work.

**Comment by shamsasari** [comment]
Back to you @exFalso

File: node-api/src/main/kotlin/net/corda/nodeapi/config/ConfigUtilities.kt (hunk 3 above)
Comment (exFalso) [inline node-api/src/main/kotlin/net/corda/nodeapi/config/ConfigUtilities.kt]: how about
```
private fun <T : Enum<T>> enumBridge(clazz: Class<T>, enumValueString: String): T {
    return java.lang.Enum.valueOf(clazz, enumValueString)
}
private class DummyEnum : Enum<DummyEnum>("", 0)

@Suppress("UNCHECKED_CAST")
private fun parseEnum(enumType: Class<*>, name: String): Enum<*> {
    return enumBridge(enumType as Class<DummyEnum>, name)
}
```

Comment (exFalso) [inline node-api/src/main/kotlin/net/corda/nodeapi/config/ConfigUtilities.kt]: actually any Enum will do fine if there are any in the std lib

Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/config/ConfigUtilities.kt]: That worked, thanks for that! Don't know what I was doing wrong before. Back to you.

**Pull request merged**